### DISCUSSION TITLE
Is there a way to specify the order of the tasks in the task panel of the TUI?

### DISCUSSION THREAD

**Discussion by Stan-Stani** [vercel/turborepo#9146] [Help] [answered]
### Summary

Alternatively, is there a way to at least set the one that's initially focused?

![explorer_0HP0Lyn0jF](https://github.com/user-attachments/assets/f0e86372-d0b8-4d61-bf89-a75aef34eda6)

**Comment by luskin** [comment]
This would be a great feature. We have over 30 tasks running and it can be a real pain sometimes to get down to the task you need to inspect.

**Comment by anthonyshew** [comment, marked as answer]
There is no way to specify an order for the task list. What order would you want to create? Additionally, are you able to guarantee that every developer in the repository would want the same order?

I would ask the same question about the initial focus as well. The task that you may believe to be the most important may not be the one that others in the repo be the most important one, so how can we assign the right initial task universally?

**Reply by anthonyshew** [reply]
Was thinking about this more. I think there's a way that we could write to disk in `.turbo` whenever you select a task in the TUI and, the next time you open the TUI, we would know to open on that task. So, if you close the TUI with `web#dev` as your last selected task, it would open again with `web#dev` selected from the start.

**Reply by traviswimer** [reply]
I would definitely prefer that solution over something like a CLI flag, since it "just works" rather than being some config you need to remember and be aware of.

Also I think a really simple improvement that could be made for the ordering of tasks would be to sort the tasks alphabetically by default. Currently, you have no way of knowing what order your tasks will be in. I think ensuring the order is at least consistent on every run would make a big difference.

**Reply by luskin** [reply]
@anthonyshew That's the exact use-case that would satisfy our needs. Right now all the engineers have to key down until they reach our nextjs web task, which sometimes can be 20+ clicks!

**Comment by jroru** [comment]
What about using the dependency topology here? In my use-case I'm interested in having the root web app dev server shown in position 1 and selected by default.

I'm sure there are cases where that's not the desire, but it seems to me that it would mostly be the case.

**Reply by anthonyshew** [reply]
https://github.com/vercel/turborepo/discussions/9146#discussioncomment-10759078

**Reply by jroru** [reply]
@anthonyshew any thoughts on using the topology instead of/as well as the proposal you linked?

**Reply by anthonyshew** [reply]
I have the same questions as I did before: What order would you want to create? Additionally, are you able to guarantee that every developer in the repository would want the same order?

The task graph itself is non-deterministic, so we can't guarantee that your task graph will evaluate into the order the same every time - which is why you see the differences that you do today. Since we can't rely on that, we need some sensible default that makes sense in all cases, but I'm not sure what that might be.

**Comment by mifopen** [comment]
> What order would you want to create?

Alphanumeric.

> are you able to guarantee that every developer in the repository would want the same order?

Yes, 100%.

The logic here is very simple, any order that:
1) is deterministic
2) is easy to scan through with your eyes
3) does not "jump"

is preferable to any other.

The only scenario when the order even matters is when you have some persistent tasks in the list (because otherwise it's anyway short-living UI that can close at any moment). It means that you have some long-running jobs that you want to look at from time to time. I cannot imagine why I would want to have any logic that contradicts to the points above.

For us this is the number one UX problem with the tool. Of course it would be nice to have a way to filter the list in process, but this is another story.

@anthonyshew, what do you think?
Would happily discuss it further. :)

PS. The suggestion does not contradict to your suggestion in the thread marked as "answer". It is useful to have some kind of "view persistence" from run to run. Same probably should be true for any other "action" that you can do in tui (e.g. hide/show sidepanel).

PPS. Of course you can have a mix of long-running and short-running tasks in the list. I do not think that it is an excuse for reordering the list in process. I still would like to check results of those short-running tasks without chasing them from top to bottom of the list. Some shortcut to hide completed tasks is good alternative in my opinion. (Could be also persisted as a TUI setting locally).

**Reply by anthonyshew** [reply]
Unfortunately, alphanumeric isn't as obvious as it may seem. We also have heard feedback that alphanumeric is clearly _not_ the right solution, and that topological ordering would be vastly preferable. While we do realize having the order be non-deterministic isn't ideal, it's still not clear what the right solution is - or that we may need to simply make a decision on an ordering strategy, accepting its faults, and go with it.

> The only scenario when the order even matters is when you have some persistent tasks in the list

This isn't correct. You can be running non-persistent tasks that take awhile to finish, and want to scroll around in the task list to take a look at how things went. I can imagine that someone doing this would want to intuit the ordering for all status groups of their tasks, not only for the in-progress ones.

> it would be nice to have a way to filter the list

You can with `/`. It appears we haven't provided a UI hint for that, which we can fix.

**Reply by traviswimer** [reply]
@anthonyshew 

> We also have heard feedback that alphanumeric is clearly not the right solution

I see your point but, would you agree that there is no "right" solution for everyone? If so, would it be feasible to provide a way to allow specifying a custom sorting function? (I'm not really familiar with Rust)

Also, would you agree that alphanumeric sorting **IS** the right solution for some users? If so, do you see any harm in adding it as an option? It might not be the best solution, but it is an improvement that I assume would be very easy to implement.

**Reply by anthonyshew** [reply]
No, I would not agree that alphanumeric is right - likely not for anyone, actually. The reason for this is that scrolling through the list of tasks would become difficult to work with. The order would be changing as task move in and out of different states, causing the options that are available to be constantly changing.

We append to a set of the tasks in a status because of this problem.

**Comment by goszczynskip** [comment]
@anthonyshew Multiple things make the current solution hard for developers to work with:
- Unpredictable ordering of tasks. Each time I run a small/mid-size monorepo project eg. create-t3-turbo  I get a different order of tasks based on task completion time. This leads to worse DX since I need to scan all the tasks to find the one I'm interested in. The problem isn't about making different ordering but rather making it predictable between runs.
- Long-running tasks even though they are kept on the top are also in "random" order without an option to control it. I understand ordering is influenced by dependent packages that long-running tasks are waiting for.
- No option to control the default selected task. Projects with multiple long-running tasks where one is the main one are left in a "weird" state. For example, create-t3-turbo with multiple apps where some are supplementary like local OTEL or proxy and some are core that should be displayed by default like Next. 

The solution to these problems is to allow the defining of a task's custom order via turbo.json and package.json

Example 1 - a single field that controls ordering behavior in turbo.json.
```
{
  "tasksOrder": "alphanumeric" | "dynamic", <- alphanumeric makes ordering static, dynamic is the current algorithm
}
```

Example 2 - more granular ordering based on nested turbo.json
```
// turbo.json
{
  "tasksOrder": "dynamic",
  "defaultOrder": 100 <- optional value, sets the order for all of the tasks without specified one
}

// apps/nextjs/turbo.json
{
  "extends": ["//"],
  "order": 0, <- overrides default order. Lower values are put on the top.
  "tuiDefault": true <- keeps the currently viewed task in tui to this task
}
```

**Reply by anthonyshew** [reply]
Have you gotten to try the latest version of Turborepo? It keeps the same task selected across runs, and sounds like it solves for the pains you're mentioning.

https://turbo.build/blog/turbo-2-4#terminal-ui-improvements

**Comment by Youpiiiii** [comment]
The problem is changing the order; we need to search within the list. A simple alphanumeric sort would be a good start.

**Reply by anthonyshew** [reply]
Answered why alphanumeric won't work out well here: https://github.com/vercel/turborepo/discussions/9146#discussioncomment-11352830

**Comment by kachurun** [comment]
I have a typical `apps + packages` structure in my monorepo, where apps are real applications (client admin panel, Storybook, website, internal admin panel, etc.), and packages are reusable across these applications (like components, libs, etc.).

I don't care about sorting in packages, and I actually like the current order where the last active is first and the completed is last.

But my apps are always persistent and active, and no matter what, I want to see them first.

For example:

```
» Storybook#watch
» components#watch
» libs#watch
⊙ components#build
⊙ libs#build

(NOTE: libs and components are duplicated probably because I'm an idiot and don't understand something, but it's the only way I get stable HMR without bugs on clean runs and without full rebuilds after any change in components or libs.)
```

So, there is no situation where I want to see components or libs above Storybook.
